In [3]:
import pandas as pd 
import numpy as np 


<hr style="border-top: 2px solid black;">

### Q1. Versión de Pandas

In [3]:
pd.__version__

'2.3.2'

In [5]:
#obtner datos 
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv

--2025-09-15 14:42:53--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 298573 (292K) [text/plain]
Saving to: ‘laptops.csv’

laptops.csv         100%[===================>] 291.58K  --.-KB/s    in 0.1s    

2025-09-15 14:42:54 (2.55 MB/s) - ‘laptops.csv’ saved [298573/298573]



In [4]:
df = pd.read_csv('laptops.csv')

In [5]:
df.columns

Index(['Laptop', 'Status', 'Brand', 'Model', 'CPU', 'RAM', 'Storage',
       'Storage type', 'GPU', 'Screen', 'Touch', 'Final Price'],
      dtype='object')

In [7]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [9]:
string =  list(df.dtypes[df.dtypes == "object"].index)
string

['laptop', 'status', 'brand', 'model', 'cpu', 'storage_type', 'gpu', 'touch']

In [10]:
for col in string:
    df[col] = df[col].str.lower().str.replace(' ', '_')

In [13]:
df.sample(5)

laptop status   brand  \
788   acer_predator_triton_500_se_pt516-51s-75k6_int...    new    acer   
1447  hp_pavilion_14-dv1009ns_intel_core_i5-1155g7/1...    new      hp   
830   asus_f515ea-ej1564w_intel_core_i3-1115g4/8gb/2...    new    asus   
677   lenovo_thinkpad_l14_gen_2_intel_core_i5-1145g7...    new  lenovo   
451   lenovo_ideapad_flex_5_14itl05_intel_core_i7-11...    new  lenovo   

         model            cpu  ram  storage storage_type       gpu  screen  \
788   predator  intel_core_i7   32     1000          ssd  rtx_3080    16.0   
1447  pavilion  intel_core_i5   16      512          ssd       NaN    14.0   
830       f515  intel_core_i3    8      256          ssd       NaN    15.6   
677   thinkpad  intel_core_i5   16      256          ssd       NaN    14.0   
451    ideapad  intel_core_i7   16      512          ssd       NaN    14.0   

     touch  final_price  
788     no      2739.00  
1447    no       920.82  
830     no       391.60  
677     no       869.00  
451    yes       979.00

<hr style="border-top: 2px solid black;">

### Q2. Recuento de registros
¿Cuántos registros hay en el conjunto de datos?
- 12
- 1000
- 2160 ✅
- 12160 


In [10]:
registros = df.shape[0]
registros

2160

<hr style="border-top: 2px solid black;">

### Q3. Marcas de portátiles
¿Cuántas marcas de computadoras portátiles se presentan en el conjunto de datos?

- 12
- 27 ✅
- 28
- 2160

In [14]:
marcas = df['Brand'].nunique()
marcas

27

<hr style="border-top: 2px solid black;">

### Q4. Valores faltantes
¿Cuántas columnas del conjunto de datos tienen valores faltantes?

- 0
- 1
- 2
- 3 ✅

In [22]:
valoresNulos = df.isnull().any().sum()
print(valoresNulos)

3


<hr style="border-top: 2px solid black;">

### Q5. Precio final máximo
¿Cuál es el precio final máximo de los portátiles Dell en el conjunto de datos?

- 869
- 3691
- 3849
- 3936 ✅

In [26]:
precio_maxixo_dell = df[df["brand"] == "dell"]["final_price"].max()
precio_maxixo_dell

3936.0

<hr style="border-top: 2px solid black;">

### Q6. Valor medio de la pantalla
Encuentra el valor mediano de Screen columna en el conjunto de datos.
A continuación, calcule el valor más frecuente del mismo Screen columna.
Uso fillna método para completar los valores faltantes en Screen columna con el valor más frecuente del paso anterior.
Ahora, calcule el valor mediano de Screen una vez más.
¿Ha cambiado?

Sugerencia: consulte existente mode y median funciones para completar la tarea.

- Sí
- No ✅

In [36]:
valorMediano = df["screen"].median()
print(f'Este es el valor mediano {valorMediano}')
valorFrecuente = df["screen"].mode()[0]
print(f'Este es el valor mas frecuente {valorFrecuente}')
valoresFaltantes = df.fillna(valorFrecuente)
valorMediano2 = df.screen.median()
print(f'Este es valor mediano sin datos faltantes {valorMediano2}')



Este es el valor mediano 15.6
Este es el valor mas frecuente 15.6
Este es valor mediano sin datos faltantes 15.6


<hr style="border-top: 2px solid black;">

### Q7. Suma de pesos
Seleccione todas las computadoras portátiles "Innjoo" del conjunto de datos.
Seleccione solo columnas RAM, Storage, Screen.
Obtenga la matriz NumPy subyacente. Vamos a llamarlo X.
Calcular la multiplicación matriz-matriz entre la transpuesta de X y X. Para obtener la transposición, utilice X.T. Llamemos al resultado XTX.
Calcule la inversa de XTX.
Crea una matriz y con valores [1100, 1300, 800, 900, 1000, 1100].
Multiplica la inversa de XTX con la transposición de X, y luego multiplica el resultado por y. Llame al resultado w.
¿Cuál es la suma de todos los elementos del resultado?
Nota: Acabas de implementar la regresión lineal. Hablaremos de ello en la próxima lección.

- 0,43
- 45,29
- 45,58
- 91.30 ✅

In [42]:
X = df[df['brand'] == 'innjoo'][['ram', 'storage', 'screen' ]]
XTX = X.T.dot(X)
XTX
XTX_inv = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800, 900, 1000, 1100])
w = XTX_inv.dot(X.T).dot(y)
print(w.sum().round(3))

91.3
